Fix #1216, Add support for spoiler formatting & attachments

Chris-Johnston opened the pull request:
Fix #1216 

Implements support for formatting text as spoilers, and spoiler attachments.
- Adds a method to convert text to spoiler text in the `Format` class.
- Adds `|` to the list of characters to escape, since this is used for formatting spoilers.
- Adds `isSpoiler` optional parameter to `SendFileAsync`.
    - Adds a field to `UploadFileParams`. When this is set to true, the file name will be prefixed with the text `"SPOILER_"`. Will also send the `hasSpoiler` value in the upload request body.
    - From my observations, the client only cares that attachments begin with `SPOILER_`, and not that the `hasSpoiler` property is supplied.
- Adds an extension method for `IAttachment` which checks if an attachment is a spoiler. Spoilers attachments can be of any type, not just images.

Chris-Johnston committed:
Add support for spoiler formatting

Chris-Johnston committed:
Implement support for sending message attachments that are spoilers

Chris-Johnston committed:
use consistent naming

Chris-Johnston committed:
update docstring for new isSpoiler param

Chris-Johnston committed:
move extension method to be under core project, make spoiler prefix a const

Chris-Johnston committed:
typo fix

Chris-Johnston committed:
update missing xmldocs

Joe4evr commented on src/Discord.Net.Core/Entities/Messages/IAttachment.cs:
```diff
@@ -5,6 +5,11 @@ namespace Discord
+        /// <summary>
+        ///     The prefix applied to files to indicate that it is a spoiler.
+        /// </summary>
+        public const string SpoilerPrefix = "SPOILER_";
```
:eyes:? We're not building with preview compiler versions, are we?
How does this not produce a CI failure?

Joe4evr replied:
[SharpLab, for reference.](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABADAAmwEYBuAWAChsBmfGAO2GjgEMBjCfASQDEB7PpQDelfGMK02fegGdghIgQAa+ALz4ARPz4byFAL5A=)

Chris-Johnston replied:
Nope, this was just my mistake. I didn't build before committing. Also, it seems that CI hasn't been working since we moved to the org.

FiniteReality replied:
Yeah, we need to get it moved over from Volt's personal repo. cc @RogueException

Chris-Johnston committed:
move SpoilerPrefix const outside of interface

AntiTcb commented:
Should also be added to `DiscordWebhookClient.SendFileAsync` if those support spoiler attachments as well.

Chris-Johnston committed:
Add isSpoiler support to webhook client

adds the isSpoiler field to uploading files with a webhook, which will only
insert "SPOILER_" to the start of the filename. This does not include other
fields in the payload, as this is not in the documentation, and was not observed
like in the regular client
